pkg/didcomm/common/service/service.go

```diff
@@ -30,6 +37,39 @@ type DIDCommMsg struct {
 	Payload  []byte
 	// TODO : might need refactor as per the issue-226
 	OutboundDestination *Destination
+	thID                *string
+}
+
+func (m *DIDCommMsg) parseID() (string, error) {
+	msg := struct {
+		ID     string           `json:"@id"`
+		Thread decorator.Thread `json:"~thread"`
+	}{}
+
+	if err := json.Unmarshal(m.Payload, &msg); err != nil {
+		return "", fmt.Errorf("cannot unmarshal @id and ~thread: error=%v", err)
+	}
+
+	if len(msg.Thread.ID) > 0 {
+		return msg.Thread.ID, nil
+	}
+
+	return msg.ID, nil
+}
+
+// ThreadID returns msg ID
+func (m *DIDCommMsg) ThreadID() (string, error) {
+	if m.thID != nil {
+		return *m.thID, nil
```

Comment:
If the message does not have an ID it is valid (no error)
I would suggest to check it, and if the ID is empty we can throw an error e.g `message does not have the ID`
@troyronda @rolsonquadras @llorllale what do you think?

Reply:
DIDExchange Invitation and exchangeRequest messages will not have threadID. Not sure throwing error for those cases will make sense. 

Possible solution
- Exclude those 2 message types for error checking in this method.
- Consumer of this API for those message types shouldn't invoke the method.

@llorllale @troyronda any other thoughts ?

Reply:
@soluchok 

> If the message does not have an ID it is valid (no error)
> I would suggest to check it, and if the ID is empty we can throw an error e.g message does not have the ID

I agree that we should throw an error if the threadID cannot be determined. This should probably be a sentinel error as well (`ErrThreadIdNotFound` or something similar).

A message's threadID can be taken from one of two sources:

* Explicit: `~thread.thid`
* Implicit: `@id` (but only if there is no `~thread.thid`)

A message with `~thread.thid` but no `@id` is illegal.

Reply:
Done!